### PR TITLE
Confirmation System Refactor/Overhaul

### PR DESCRIPTION
**Pull request by suneettipirneni**
#### Description: 

This PR aims to overhaul the existing confirmation system. To illustrate the significance of this PR I'll go over how confirmations are currently handled.

### Current Way of Adding Confirmations
1. You need to add a Type to the `ConfirmationType` enum:
```java
public enum ConfirmationType {
	TOWN_DELETE,
	NATION_DELETE,
	...
	GROUP_TOGGLE_ACTION,
        MY_NEW_TYPE // Added type
}
```
2. Once the type is added you need to define how its added in `ConfirmationHandler`:
```java
public static void addConfirmation(final Resident r, final ConfirmationType type, Object extra) throws TownyException {
	// We use "extra" in certain instances like the number of days for something e.t.c
	switch (type) {
		case TOWN_DELETE:
				...
		case NATION_DELETE:
				...
		case GROUP_TOGGLE_ACTION:
				...
                  case MY_NEW_TYPE: // Added here
                        r.setConfirmationType(type);
			grouptoggleconfirmations.put(r, extra);

			new BukkitRunnable() {
				@Override
				public void run() {
					removeConfirmation(r, type, false);
				}
			}.runTaskLater(plugin, 400);
			break;
		}
	}
...
```
3. Now you need to define how this confirmation is handled in the same file:
```java
public static void handleConfirmation(Resident r, ConfirmationType type) throws TownyException {
	if (type == ConfirmationType.MY_NEW_TYPE) {
			       ...
			}
		}
	}
}
```
4.) Now you need to define the way it's cancelled
```java
public static void removeConfirmation(Resident r, ConfirmationType type, boolean successful) {
	switch (type) {
		case MY_NEW_TYPE:
		      ...
        ...
        }
}
```
5. Finally, you need to add the object in the appropriate area in the code:
```java
Resident resident = new Resident("Bob");
Object extra = new Object();
ConfirmationHandler.addConfirmation(resident, ConfirmationType.MY_NEW_TYPE, extra);
```

As you can see this process is extremely monotonous. Now let's see how the new code works

### New Way of Adding Confirmations

1. Create `Confirmation` object:
```java
Confirmation confirmation = new Confirmation(()-> System.out.println("I was confirmed!"));
```
2. Send confirmation:
```java
ConfirmationHandler.sendConfirmation(player, confirmation);
```
In addition, the confirmation message text will be sent by calling `sendConfirmation()`, you can define the message title by using `Confirmation.setTitle()`.

That's it, much less work than the first steps.

Now because the static `ConfirmationType` enum is gone, plugin developers can define and call their own confirmations. Also since the confirmations are keyed by `CommandSender` you don't have to worry about differences between players and the console.

In other statistics, because of the reduction of code needed for confirmations, the codebase actually shrinks with this PR. The codebase will trim over **300** lines of code! (nice) 

____
- [x] I have tested this pull request for defects on a server. 

By making this pull request, I represent that I have the right to waive copyright and related rights to my contribution, and agree that all copyright and related rights in my contributions are waived, and I acknowledge that the TownyAdvanced organization has the copyright to use and modify my contribution under the Towny [License](https://github.com/LlmDl/Towny/blob/master/LICENSE.md) for perpetuity.

### CODE REVIEW

File: src/com/palmergames/bukkit/towny/command/TownyAdminCommand.java [hunk 1]
```diff
@@ -1592,18 +1603,62 @@ public void purge(String[] split) {
...
+				Confirmation confirmation = new Confirmation(player);
+				confirmation.setHandler(purgeHandler);
+				ConfirmationHandler.registerConfirmation(confirmation);
+				TownyMessaging.sendConfirmationMessage(player, null, null, null, null);
 			}
 		} else { // isConsole
-			ConfirmationHandler.addConfirmation(ConfirmationType.PURGE, days);
+			Confirmation confirmation = new Confirmation(sender);
+
+			final AtomicReference<String>[] finalDays = new AtomicReference[]{new AtomicReference<>(days[0])};
```
Comment (silverwolfg11) [inline src/com/palmergames/bukkit/towny/command/TownyAdminCommand.java:1649]: Why an atomic reference?

File: src/com/palmergames/bukkit/towny/command/NationCommand.java [hunk 2]
```diff
@@ -1350,7 +1352,20 @@ public void mergeNation(Player player, String name) throws TownyException {
 				throw new TownyException(String.format(TownySettings.getLangString("msg_err_king_of_that_nation_is_not_online"), name, king.getName()));
 			}
 			TownyMessaging.sendMessage(BukkitTools.getPlayer(king.getName()), String.format(TownySettings.getLangString("msg_would_you_merge_your_nation_into_other_nation"), nation, remainingNation, remainingNation));
-			ConfirmationHandler.addConfirmation(king, ConfirmationType.NATION_MERGE, remainingNation);
+			Confirmation confirmation = new Confirmation(player);
+			Nation finalRemainingNation = remainingNation;
+			confirmation.setHandler(() -> {
+				try {
+					Nation succumbingNation = resident.getTown().getNation();
+					Nation prevailingNation = finalRemainingNation;
```
Comment (LlmDl) [inline src/com/palmergames/bukkit/towny/command/NationCommand.java:1360]: Can we not reuse the above nation and remainingNation here now?

Also finalRemainingNation is basically unused.

File: src/com/palmergames/bukkit/towny/command/PlotCommand.java [hunk 3]
```diff
@@ -400,8 +412,17 @@ public boolean parsePlotCommand(Player player, String[] split) throws TownyExcep
 									new PlotClaim(plugin, player, resident, selection, false, false, false).start();
 									continue;
 								}
+								
+								// Create confirmation.
+								Confirmation confirmation = new Confirmation(player);
+								List<WorldCoord> finalSelection = selection;
```
Comment (LlmDl) [inline src/com/palmergames/bukkit/towny/command/PlotCommand.java:418]: This selection will only be one townblock and needs to be populated with the PlotGroup's townblocks like the group claiming confirmation above.

File: src/com/palmergames/bukkit/towny/command/TownyAdminCommand.java [hunk 4]
```diff
@@ -1592,18 +1603,62 @@ public void purge(String[] split) {
...
-			
-			if (resident != null) {
-				try {
-					ConfirmationHandler.addConfirmation(resident, ConfirmationType.PURGE, days); // It takes the senders town & nation, an admin deleting another town has no confirmation.
-					TownyMessaging.sendConfirmationMessage(player, null, null, null, null);
 
-				} catch (TownyException e) {
-					TownyMessaging.sendErrorMsg(player, e.getMessage());
+			Resident finalResident = resident;
+			String finalDays = days[0];
```
Comment (LlmDl) [inline src/com/palmergames/bukkit/towny/command/TownyAdminCommand.java:1608]: Not sure why you've changed the above days to a String[] to make days[0], to create a new finalDays String to do the same logic below as was used previously.

**Pull request merged**